linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameCompressor.scala

```diff
@@ -0,0 +1,406 @@
+import org.scalajs.linker.backend.javascript.{Trees => js}
+import org.scalajs.linker.standard.{LinkedClass, ModuleSet}
+
+import org.scalajs.logging.Logger
+
+import EmitterNames._
+import PolyfillableBuiltin._
+import Transients._
+
+private[emitter] final class NameCompressor(val config: Emitter.Config) {
+  import NameCompressor._
+
+  private var entries: EntryMap = null
+  private var ancestorEntries: mutable.AnyRefMap[ClassName, AncestorNameEntry] = null
+  private var _dangerousGlobalRefs: Set[String] = null
+
+  def startRun(moduleSet: ModuleSet, logger: Logger): Unit = {
+    val traverser = new Traverser(config)
+
+    logger.time("Name compressor: Collect names") {
+      traverser.traverseModuleSet(moduleSet)
+    }
+
+    entries = traverser.entries
+    ancestorEntries = traverser.ancestorEntries
+    _dangerousGlobalRefs = traverser.dangerousGlobalRefs.toSet
+
+    logger.time("Name compressor: Allocate property names") {
+      allocatePropertyNames(traverser.propertyNamesToAvoid)
+    }
+
+    logger.time("Name compressor: Allocate ancestor names") {
+      allocateAncestorNames()
+    }
+  }
+
+  def dangerousGlobalRefs: Set[String] =
+    _dangerousGlobalRefs
+
+  def allocatedFor(className: ClassName, fieldName: FieldName): String =
+    entries((className, fieldName)).allocatedName
+
+  def allocatedFor(methodName: MethodName): String =
+    entries(methodName).allocatedName
+
+  def allocatedForAncestor(ancestor: ClassName): String =
+    ancestorEntries(ancestor).allocatedName
+
+  private def allocatePropertyNames(namesToAvoid: collection.Set[String]): Unit = {
+    val orderedEntries = entries.values.toArray
+    java.util.Arrays.sort(orderedEntries, PropertyNameEntryComparator)
+
+    val generator = new NameGenerator(namesToAvoid)
+
+    for (entry <- orderedEntries)
+      entry.allocatedName = generator.nextString()
+  }
+
+  private def allocateAncestorNames(): Unit = {
+    val orderedEntries = ancestorEntries.values.toArray
+    java.util.Arrays.sort(orderedEntries, AncestorNameEntryComparator)
+
+    val generator = new NameGenerator(BasePropertyNamesToAvoid)
+
+    for (entry <- orderedEntries)
+      entry.allocatedName = generator.nextString()
+  }
+}
+
+private[emitter] object NameCompressor {
+  /** Base set of names that should be avoided when allocating property names.
+   *
+   *  These are the reserved JS identifier names and the `then` name.
+   *  `then` is used to identify `Thenable`s, so it must not be used in any
+   *  generated situation.
+   */
+  private val BasePropertyNamesToAvoid: Set[String] =
+    NameGen.ReservedJSIdentifierNames + "then"
+
+  /** Keys of this map are `(ClassName, FieldName) | MethodName`. */
+  private type EntryMap = mutable.AnyRefMap[AnyRef, PropertyNameEntry]
+
+  private sealed abstract class PropertyNameEntry {
+    var occurrences: Int = 0
+    var allocatedName: String = null
+
+    def incOccurrences(): Unit =
+      occurrences += 1
+  }
```

Comment:
FWIW: I feel this might be a bit more mutability than necessary. But let's table that until we've figured out how we want to collect the names.